# Accessory files necessary for TnSeq analysis

## File dependencies (not generated by scripts + sequencing reads):
### External (from other research)r:
BYxRM_GenoData.csv - converted to csv from tsv from http://genomics-pubs.princeton.edu/YeastCross_BYxRM/data.shtml (Bloom et al. 2013)<br>
Clones_for_Tn96_Experiment.csv - derived from elife-27167-supp2-v2.csv with information about drug resistance added by me (Jerison et al. 2017)<br>
S288C bowtie genome build, from SGD, https://www.yeastgenome.org/strain/S288C <br>

### Internal (We made them):
Demultiplex maps, replicate-library information files<br>
Segregant exclusion files<br>
Tn96_edges_chosen_final.csv - Edges chosen for TP based on old analysis from Sergey and their s values, I have a write-up of how this was done that can be included in the supplement<br>
TnCS_Final_Rearray_Data.csv - information on the original rearray<br>
GC_cell_count_estimates.csv - data from coulter counter
